cmd/thv-operator/pkg/sources/git.go

```diff
@@ -0,0 +1,190 @@
+func NewGitSourceHandler() *GitSourceHandler {
+	return &GitSourceHandler{
+		gitClient: git.NewDefaultGitClient(),
+		validator: NewSourceDataValidator(),
+	}
+}
+
+// Validate validates the Git source configuration
+func (*GitSourceHandler) Validate(source *mcpv1alpha1.MCPRegistrySource) error {
+	if source.Type != mcpv1alpha1.RegistrySourceTypeGit {
+		return fmt.Errorf("invalid source type: expected %s, got %s",
+			mcpv1alpha1.RegistrySourceTypeGit, source.Type)
+	}
+
+	if source.Git == nil {
+		return fmt.Errorf("git configuration is required for source type %s",
+			mcpv1alpha1.RegistrySourceTypeGit)
+	}
+
+	gitSource := source.Git
+
+	if gitSource.Repository == "" {
+		return fmt.Errorf("git repository URL cannot be empty")
+	}
+
+	// Validate mutually exclusive branch/tag/commit
+	specified := 0
+	if gitSource.Branch != "" {
+		specified++
+	}
+	if gitSource.Tag != "" {
+		specified++
+	}
+	if gitSource.Commit != "" {
+		specified++
+	}
+
+	if specified > 1 {
+		return fmt.Errorf("only one of branch, tag, or commit may be specified")
+	}
+
+	// Set default path if not specified
+	if gitSource.Path == "" {
+		gitSource.Path = DefaultRegistryDataFile
+	}
+
+	return nil
+}
+
+// FetchRegistry retrieves registry data from the Git repository
+func (h *GitSourceHandler) FetchRegistry(ctx context.Context, mcpRegistry *mcpv1alpha1.MCPRegistry) (*FetchResult, error) {
+
+	// Validate source configuration
+	if err := h.Validate(&mcpRegistry.Spec.Source); err != nil {
+		return nil, fmt.Errorf("source validation failed: %w", err)
+	}
+
+	// Create temporary directory for cloning
+	tempDir, err := os.MkdirTemp("", "mcpregistry-git-*")
+	if err != nil {
+		return nil, fmt.Errorf("failed to create temporary directory: %w", err)
+	}
+	defer os.RemoveAll(tempDir)
+
+	// Prepare clone configuration
+	cloneConfig := &git.CloneConfig{
+		URL:       mcpRegistry.Spec.Source.Git.Repository,
+		Branch:    mcpRegistry.Spec.Source.Git.Branch,
+		Tag:       mcpRegistry.Spec.Source.Git.Tag,
+		Commit:    mcpRegistry.Spec.Source.Git.Commit,
+		Directory: tempDir,
+	}
+
+	// Clone the repository
+	repoInfo, err := h.gitClient.Clone(ctx, cloneConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to clone repository: %w", err)
+	}
+
+	// Ensure cleanup
+	defer func() {
+		if cleanupErr := h.gitClient.Cleanup(repoInfo); cleanupErr != nil {
+			// Log error but don't fail the operation
+			fmt.Printf("Warning: failed to cleanup repository: %v\n", cleanupErr)
+		}
+	}()
+
+	// Get file content from repository
+	filePath := mcpRegistry.Spec.Source.Git.Path
+	if filePath == "" {
+		filePath = DefaultRegistryDataFile
+	}
+
+	registryData, err := h.gitClient.GetFileContent(repoInfo, filePath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get file %s from repository: %w", filePath, err)
+	}
+
+	// Validate and parse registry data
+	reg, err := h.validator.ValidateData(registryData, mcpRegistry.Spec.Source.Format)
```

Comment:
It seems wasteful to do a full clone of the repo again. Could we split ValidateData into e.g. `validateDataFromRepoInfo` and let only the public `ValidateData` clone again. If we want to do more optimization we could later on as well.

Reply:
(I hope I am reading the code right)

When the controller calls `reconcileSync` it first calls `ShouldSync` which calls `IsDataChanged` -> `CurrentHash` -> `Clone`.

Later the controller calls `PerformSync` -> `fetchAndProcessRegistryData` -> `FetchRegistry` -> `Clone`.

So we have 2 clones per reconcile. We don't have to solve this now for MVP-1.

It's just that I've been hit by cloning large repos in past projects.